File: backend/app/core/tag_policy.py
from __future__ import annotations


def normalize_tags(tags: list[str] | None) -> list[str]:
    if not tags:
        return []
    seen: set[str] = set()
    result: list[str] = []
    for raw in tags:
        cleaned = " ".join(str(raw).split())
        if not cleaned or cleaned in seen:
            continue
        seen.add(cleaned)
        result.append(cleaned)
    return result

File: frontend/src/services/tagPolicy.ts
export const TAG_GENERATED = "AI 生成";
export const TAG_DOCUMENTS = "文档资料";
export const TAG_IMPORTED = "导入内容";
export const TAG_YOUTUBE = "视频";

export const SYSTEM_TAGS: readonly string[] = [
  TAG_GENERATED,
  TAG_DOCUMENTS,
  TAG_IMPORTED,
  TAG_YOUTUBE,
];

export function isSystemTag(tag: string): boolean {
  return SYSTEM_TAGS.includes(tag);
}

export function normalizeTags(tags: readonly string[] | null | undefined): string[] {
  if (!tags) return [];
  const seen = new Set<string>();
  const result: string[] = [];
  for (const raw of tags) {
    const cleaned = String(raw).split(/\s+/).filter(Boolean).join(" ");
    if (!cleaned || seen.has(cleaned)) continue;
    seen.add(cleaned);
    result.push(cleaned);
  }
  return result;
}
